### PR TITLE
v3.0.0 - Open Insights

### PR DESCRIPTION
**Pull request by phamann**
### TL;DR
This is a large refactor of the library to use [Open Insights](https://github.com/openinsights/openinsights) as the core runtime and the associated [Open Insights Fastly provider](https://github.com/fastly/open-insights-provider-fastly) for the Fastly business logic. 

While a big change, there is no functional difference to v2, the main difference being that we we have moved the logic form this repository into the respective Open Insights repository and re-imported it as dependencies of this project. As Open Insights was inspired by this project :smile:. Hence why the diff is mainly deletions.

This all comes together and will make more sense reading: [`src/index.ts`](https://github.com/fastly/insights.js/pull/62/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80).

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -12,39 +17,50 @@ declare global {
...
+    report_errors: true,
+    token: params.token,
+  };
+
+  settingsBuilder.addProvider(new FastlyProvider(settings));
+
+  init(settingsBuilder.toSettings()).then((result) => {
+    // We use apply here as we don't want to directly assign i.e.
+    // replace the existing results array.
+    Array.prototype.push.apply(state.results, result.testResults);
```
Comment (doramatadora) [inline src/index.ts:62]: Oh, I haven't come across `Array.prototype.push.apply` in a while :) Are we using this to appease the linter? If so, I'd like to offer a nit-picky suggestion to save some code comments and/or improve readability:
```suggestion
    state.results.push(...result.testResults);
```
The targets in `.browserlistrc` mean that this gets transpiled to `state.results.apply(state.results, result.testResults)` anyway, which is equivalent to `Array.prototype.push.apply(state.results, result.testResults);`. Untranspiled (ESM version?), `arr.push(...extraElements)` [is marginally faster than](https://www.measurethat.net/Benchmarks/Show/2279/0/arrayprototypeconcat-vs-spread-operator-vs-apply) `Apply.prototype.push.apply(arr,extraElements)`.

File: src/util/object.ts [hunk 2]
```diff
@@ -31,6 +29,7 @@ export function hasProperties(obj: any, properties: string[]): boolean {
   try {
     hasProps = properties.every((prop): boolean => hasProperty(obj, prop));
   } catch (e) {
+    console.log("inside");
```
Comment (doramatadora) [inline src/util/object.ts:32]:
```suggestion
```

File: src/util/scriptQueryParameters.ts [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import { QueryParameters } from "../@types";
+
 const getElements = (element: string): Element[] =>
   [].slice.call(document.getElementsByTagName(element));
```
Comment (doramatadora) [inline src/util/scriptQueryParameters.ts:4]: Another nit-picky suggestion for readability:
```suggestion
  [...document.getElementsByTagName(element)];
```

**Review by doramatadora — APPROVED** [review]
Looks great! Towards the end of my review, I came across `tsconfig` and a target setting to `ES5`, which makes my smart Alec comments about transpile targets redundant 😅  My suggestions are mostly nit-picky, in the interest of not mixing ES6 syntax with older JS, for readability.

File: LICENSE_TEMPLATE [hunk 4]
```diff
@@ -0,0 +1,24 @@
+Fastly Insights.js
```
[the remark this answers is not in the document]
Comment (phamann) [reply LICENSE_TEMPLATE:1]: This was agreed upon by our legal department and wouldn't want the template diverging from the real LICENSE file in the repo.

File: src/index.ts (hunk 1 above)
Comment (phamann) [reply src/index.ts:62]: Apologies, the comment above this line should be more descriptive as to why we **HAVE** to use `Array.prototype.push.apply()` in this instance. 

For the https://insights.fastlylabs.com website we have a demo application which loads insights.js via a script tag and registers a `new Proxy()` on the window global `FASTLY.results`, so that we can spy on the results without manipulating the third-party script. 

TL;DR you use `Array.prototype.push.apply()` when you're interested in using `.push()` with an object thats isn't really an Array.

File: src/util/scriptQueryParameters.ts (hunk 3 above)
Comment (phamann) [reply src/util/scriptQueryParameters.ts:4]: This function was previous used outside of the feature detection and thus needed to be bullet proof. While I agree the typescript compiler will handle this for us, if you compare the ES% output results for the two options in the playground you'll see why `slice.call()` is still preferred :)

File: src/util/object.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/* eslint "@typescript-eslint/explicit-module-boundary-types": off */
+
+// Utility function to test whether an object has own nested property
+// This is used instead of Object.prototype.hasOwnProperty() due to browser support
+export function hasProperty(obj: any, prop: string): boolean {
```
Comment (psbanka) [inline src/util/object.ts]: out of curiosity, wouldn't it be better to use js-doc blocks? i.e.:
```javascript
/**
 * Utility function to test whether an object has own nested property
 * This is used instead of Object.prototype.hasOwnProperty() due to browser support
 **/
export function hasProperty(obj: any, prop: string): boolean {
```

**Review by psbanka — APPROVED** [review]
Looks great! A few comments that you can take or leave
